### ISSUE TITLE
Liquidsoap overruns with two harbor inputs

### ISSUE THREAD

**Issue by tatokis** [savonet/liquidsoap#184]
Running liquidsoap with two harbor inputs, causes buffer overruns and it takes some time to switch to the other harbor input.
Once it does, it is quite choppy and sometimes just stops and falls back to the fallback harbor stream. Moreover, we start getting overruns on the other stream after some time.
That can be seen when looking at the log.
This is a DJ (live) and 24/7 stream setup. What we are trying to do right now is, have an external 24/7 feed from another server getting streamed to the 24/7 harbor input and when a DJ goes live, on the live harbor input, have liquidsoap switch to the live input, while still having the 24/7 stream connected to the other harbor input.
There is no file input, just two input.harbor
Also tried with one harbor and one http input but the same behaviour was noticed.
Overruns occur even with silence from mksafe() on 247.ogg
Here is a part of the log.

```
2014/07/11 15:50:14 [live:3] Buffer overrun: Dropping 0.02s.
2014/07/11 15:50:14 [live:3] Buffer overrun: Dropping 0.02s.
2014/07/11 15:50:14 [live:3] Buffer overrun: Dropping 0.02s.
2014/07/11 15:50:14 [live:3] Buffer overrun: Dropping 0.02s.
2014/07/11 15:50:14 [live:3] Buffer overrun: Dropping 0.02s.
2014/07/11 15:50:14 [live:3] Buffer overrun: Dropping 0.02s.
2014/07/11 15:50:14 [live:3] Buffer overrun: Dropping 0.02s.
2014/07/11 15:50:14 [live:3] Buffer overrun: Dropping 0.02s.
2014/07/11 15:50:14 [live:3] Buffer overrun: Dropping 0.02s.
2014/07/11 15:50:14 [live:3] Buffer overrun: Dropping 0.02s.
```

I tried renicing liquidsoap to -10 but that did not fix anything.
I also tested this on my local PC which has an AMD FX 8350 CPU on it, just to rule out that it is the hardware, and the same behaviour was repeated.
I also tried disabling the multiple outputs and leaving only one on but again, same thing.
Here is our current config

```
#Log
set("log.file.path","/tmp/liquidsoap.log")
#Harbor setup
set("harbor.bind_addr","0.0.0.0")
set("harbor.verbose",true)
set("log.level", 4)
#Harbor Live stream
live = input.harbor("live.ogg",port=8002,password="hackme",id="live")
#24/7 stream
input247 = mksafe(input.harbor("247.ogg",port=8002,password="hackme",id="247"))
# fallback
input = fallback([live,input247])

#Outputs
output.icecast(
     %opus(application="audio", channels=2, bitrate=56, vbr="constrained", signal="music"),
     host = "localhost",
     port = 8000,
     password = "hackme",
     mount = "/opustest.ogg",
     name = "RadioSEGA",
     description = "RadioSEGA Test Stream",
     url = "http://www.radiosega.net",
     input
)
#...
```

I only pasted the opus output config, there are others too but I do not think it affects the inputs.
Operating system on the server is Ubuntu Server 12.04 LTS and the PC I tested it on, runs Ubuntu Desktop 14.04

**Comment by leofiore** [comment]
i can confirm the problem, although on a different scenario. I have a liquidsoap instance running on a computer and idjc as harbour input running on the same machine. IDJC is used for about 6 hours a day for live broadcasting. 

the IDJC stream is a mp3 128kbps CBR stream with a 44100 samplerate.

The buffer overruns occurs anytime there is a network problem with the icecast server. The connection is not very reliable and sometimes liquidsoap needs to drop audio packages on the ouptut stream. If the output skipping occurs more than once, from time to time the harbour input is dropped and the harbour source is disconnected. This behaviour is repeated for at least 3 or 4 seconds.

IMO the problem is related to how the network sockets and buffers are handled by ocaml, but i have no experience about the language and the framework

**Comment by dtcooper** [comment]
Experiencing the same problem. Happy to help, but not really sure where to start.

**Comment by tatokis** [comment]
After a long break, I finally figured out the solution to my issue (and forgot to post it).
Turns out, there were two issues with the config.

The main one, causing the overruns is that once the second stream took over the output, the first one had nowhere to go, so there was a need to set up a dummy output for it.


The second issue was that the fallback was track sensitive, which caused that delay in switching over.

I have attached the config that works for me.
While there are some issues with it, I hope it can help you.

[liquidsoapconfig.liq.txt](https://github.com/savonet/liquidsoap/files/601880/liquidsoapconfig.liq.txt)
